File: ios/Classes/WearablesMethodCallHandler.swift
import Flutter
import MWDATCore
import MWDATCamera

class WearablesMethodCallHandler: NSObject {
    private let streamStateHandler: StreamStateStreamHandler
    private let videoFrameInfoHandler: VideoFrameInfoStreamHandler
    private let streamErrorsHandler: StreamErrorsStreamHandler
    private var textureRegistry: FlutterTextureRegistry?

    private var streamSessionManager: StreamSessionManager?
    private let mockDeviceKitHandler = MockDeviceKitHandler()
    private var wearables: WearablesInterface?

    init(
        streamStateHandler: StreamStateStreamHandler,
        videoFrameInfoHandler: VideoFrameInfoStreamHandler,
        streamErrorsHandler: StreamErrorsStreamHandler
    ) {
        self.streamStateHandler = streamStateHandler
        self.videoFrameInfoHandler = videoFrameInfoHandler
        self.streamErrorsHandler = streamErrorsHandler
        super.init()
    }

    func setTextureRegistry(_ registry: FlutterTextureRegistry) {
        self.textureRegistry = registry
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any]

        switch call.method {
        case "initialize":
            handleInitialize(result: result)
        case "startRegistration":
            handleStartRegistration(result: result)
        case "startUnregistration":
            handleStartUnregistration(result: result)
        case "checkPermissionStatus":
            handleCheckPermissionStatus(args: args, result: result)
        case "requestPermission":
            handleRequestPermission(args: args, result: result)
        case "handleUrl":
            handleUrl(args: args, result: result)
        case "startStreamSession":
            handleStartStreamSession(args: args, result: result)
        case "stopStreamSession":
            handleStopStreamSession(result: result)
        case "capturePhoto":
            handleCapturePhoto(result: result)
        case "dispose":
            handleDispose(result: result)
        case "mock_pairRaybanMeta":
            mockDeviceKitHandler.pairRaybanMeta(result: result)
        case "mock_unpairDevice":
            let deviceId = args?["deviceId"] as? String ?? ""
            mockDeviceKitHandler.unpairDevice(deviceId: deviceId, result: result)
        case "mock_deviceAction":
            let deviceId = args?["deviceId"] as? String ?? ""
            let action = args?["action"] as? String ?? ""
            mockDeviceKitHandler.deviceAction(deviceId: deviceId, action: action, result: result)
        case "mock_setCameraFeed":
            let deviceId = args?["deviceId"] as? String ?? ""
            let filePath = args?["filePath"] as? String ?? ""
            mockDeviceKitHandler.setCameraFeed(deviceId: deviceId, filePath: filePath, result: result)
        case "mock_setCapturedImage":
            let deviceId = args?["deviceId"] as? String ?? ""
            let filePath = args?["filePath"] as? String ?? ""
            mockDeviceKitHandler.setCapturedImage(deviceId: deviceId, filePath: filePath, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func handleInitialize(result: @escaping FlutterResult) {
        do {
            try Wearables.configure()
            wearables = Wearables.shared
            result(nil)
        } catch {
            result(FlutterError(code: "INIT_ERROR", message: "Failed to initialize: \(error)", details: nil))
        }
    }

    private func handleStartRegistration(result: @escaping FlutterResult) {
        guard let wearables = wearables else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "SDK not initialized", details: nil))
            return
        }
        Task { @MainActor in
            do {
                try await wearables.startRegistration()
                result(nil)
            } catch {
                result(FlutterError(code: "REGISTRATION_ERROR", message: "Failed to start registration: \(error)", details: nil))
            }
        }
    }

    private func handleStartUnregistration(result: @escaping FlutterResult) {
        guard let wearables = wearables else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "SDK not initialized", details: nil))
            return
        }
        Task { @MainActor in
            do {
                try await wearables.startUnregistration()
                result(nil)
            } catch {
                result(FlutterError(code: "UNREGISTRATION_ERROR", message: "Failed to unregister: \(error)", details: nil))
            }
        }
    }

    private func handleCheckPermissionStatus(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let wearables = wearables else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "SDK not initialized", details: nil))
            return
        }
        Task { @MainActor in
            do {
                let status = try await wearables.checkPermissionStatus(.camera)
                result(Converters.permissionStatusToString(status))
            } catch {
                result(FlutterError(code: "PERMISSION_ERROR", message: "Permission check failed: \(error)", details: nil))
            }
        }
    }

    private func handleRequestPermission(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let wearables = wearables else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "SDK not initialized", details: nil))
            return
        }
        Task { @MainActor in
            do {
                let status = try await wearables.requestPermission(.camera)
                result(Converters.permissionStatusToString(status))
            } catch {
                result(FlutterError(code: "PERMISSION_ERROR", message: "Permission request failed: \(error)", details: nil))
            }
        }
    }

    private func handleUrl(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let urlString = args?["url"] as? String,
              let url = URL(string: urlString),
              let wearables = wearables else {
            result(false)
            return
        }
        let handled = wearables.handleUrl(url)
        result(handled)
    }

    private func handleStartStreamSession(args: [String: Any]?, result: @escaping FlutterResult) {
        guard let wearables = wearables else {
            result(FlutterError(code: "NOT_INITIALIZED", message: "SDK not initialized", details: nil))
            return
        }
        guard let registry = textureRegistry else {
            result(FlutterError(code: "NO_REGISTRY", message: "Texture registry not available", details: nil))
            return
        }

        let resolutionStr = args?["videoQuality"] as? String ?? "medium"
        let frameRate = args?["frameRate"] as? Int ?? 24
        let resolution = Converters.streamingResolutionFromString(resolutionStr)

        let manager = StreamSessionManager(
            textureRegistry: registry,
            streamStateHandler: streamStateHandler,
            videoFrameInfoHandler: videoFrameInfoHandler,
            streamErrorsHandler: streamErrorsHandler
        )
        streamSessionManager = manager

        let sessionInfo = manager.startSession(resolution: resolution, frameRate: frameRate, wearables: wearables)
        result(sessionInfo)
    }

    private func handleStopStreamSession(result: @escaping FlutterResult) {
        streamSessionManager?.stopSession()
        streamSessionManager = nil
        result(nil)
    }

    private func handleCapturePhoto(result: @escaping FlutterResult) {
        guard let manager = streamSessionManager else {
            result(FlutterError(code: "NO_SESSION", message: "No active stream session", details: nil))
            return
        }
        manager.capturePhoto { photoResult in
            switch photoResult {
            case .success(let data):
                result(data)
            case .failure(let error):
                result(FlutterError(code: "CAPTURE_ERROR", message: "Photo capture failed: \(error.localizedDescription)", details: nil))
            }
        }
    }

    private func handleDispose(result: @escaping FlutterResult) {
        streamSessionManager?.stopSession()
        streamSessionManager = nil
        mockDeviceKitHandler.dispose()
        result(nil)
    }

    func dispose() {
        streamSessionManager?.stopSession()
        streamSessionManager = nil
        mockDeviceKitHandler.dispose()
    }
}

File: android/src/main/kotlin/com/meta/wearable/flutter/WearablesMethodCallHandler.kt
package com.meta.wearable.flutter

import android.app.Activity
import android.app.Application
import com.meta.wearable.dat.camera.types.VideoQuality
import com.meta.wearable.dat.core.Wearables
import com.meta.wearable.dat.core.types.DatResult
import com.meta.wearable.dat.core.types.Permission
import com.meta.wearable.dat.core.types.PermissionError
import com.meta.wearable.dat.core.types.PermissionStatus
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.view.TextureRegistry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class WearablesMethodCallHandler(
    private val scope: CoroutineScope,
    private val streamStateHandler: WearablesStreamHandlers.StreamStateHandler,
    private val videoFrameInfoHandler: WearablesStreamHandlers.VideoFrameInfoHandler,
    private val streamErrorsHandler: WearablesStreamHandlers.StreamErrorsHandler
) : MethodChannel.MethodCallHandler {

    companion object {
        const val PERMISSION_REQUEST_CODE = 48291
    }

    var activity: Activity? = null
    var application: Application? = null
    var textureRegistry: TextureRegistry? = null
    var onInitialized: (() -> Unit)? = null
    var permissionContract: Wearables.RequestPermissionContract? = null

    private var streamSessionManager: StreamSessionManager? = null
    private var mockDeviceKitHandler: MockDeviceKitHandler? = null
    private var pendingPermissionResult: MethodChannel.Result? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> handleInitialize(result)
            "startRegistration" -> handleStartRegistration(result)
            "startUnregistration" -> handleStartUnregistration(result)
            "checkPermissionStatus" -> handleCheckPermissionStatus(call, result)
            "requestPermission" -> handleRequestPermission(call, result)
            "handleUrl" -> handleUrl(call, result)
            "startStreamSession" -> handleStartStreamSession(call, result)
            "stopStreamSession" -> handleStopStreamSession(result)
            "capturePhoto" -> handleCapturePhoto(result)
            "dispose" -> handleDispose(result)
            "mock_pairRaybanMeta" -> getMockHandler()?.pairRaybanMeta(result)
                ?: result.error("NO_APP", "Application not available", null)
            "mock_unpairDevice" -> {
                val deviceId = call.argument<String>("deviceId")!!
                getMockHandler()?.unpairDevice(deviceId, result)
                    ?: result.error("NO_APP", "Application not available", null)
            }
            "mock_deviceAction" -> {
                val deviceId = call.argument<String>("deviceId")!!
                val action = call.argument<String>("action")!!
                getMockHandler()?.deviceAction(deviceId, action, result)
                    ?: result.error("NO_APP", "Application not available", null)
            }
            "mock_setCameraFeed" -> {
                val deviceId = call.argument<String>("deviceId")!!
                val filePath = call.argument<String>("filePath")!!
                getMockHandler()?.setCameraFeed(deviceId, filePath, result)
                    ?: result.error("NO_APP", "Application not available", null)
            }
            "mock_setCapturedImage" -> {
                val deviceId = call.argument<String>("deviceId")!!
                val filePath = call.argument<String>("filePath")!!
                getMockHandler()?.setCapturedImage(deviceId, filePath, result)
                    ?: result.error("NO_APP", "Application not available", null)
            }
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(result: MethodChannel.Result) {
        val app = application
        if (app == null) {
            result.error("NO_CONTEXT", "Application context not available", null)
            return
        }
        try {
            Wearables.initialize(app)
            onInitialized?.invoke()
            result.success(null)
        } catch (e: Exception) {
            result.error("INIT_ERROR", "Failed to initialize: ${e.message}", null)
        }
    }

    private fun handleStartRegistration(result: MethodChannel.Result) {
        val act = activity
        if (act == null) {
            result.error("NO_ACTIVITY", "Activity not available for registration", null)
            return
        }
        try {
            Wearables.startRegistration(act)
            result.success(null)
        } catch (e: Exception) {
            result.error("REGISTRATION_ERROR", "Failed to start registration: ${e.message}", null)
        }
    }

    private fun handleStartUnregistration(result: MethodChannel.Result) {
        val act = activity
        if (act == null) {
            result.error("NO_ACTIVITY", "Activity not available for unregistration", null)
            return
        }
        try {
            Wearables.startUnregistration(act)
            result.success(null)
        } catch (e: Exception) {
            result.error("UNREGISTRATION_ERROR", "Failed to start unregistration: ${e.message}", null)
        }
    }

    private fun handleCheckPermissionStatus(call: MethodCall, result: MethodChannel.Result) {
        val permissionName = call.argument<String>("permission") ?: "camera"
        scope.launch(Dispatchers.Main) {
            try {
                val permission = when (permissionName) {
                    "camera" -> Permission.CAMERA
                    else -> Permission.CAMERA
                }
                val checkResult = Wearables.checkPermissionStatus(permission)
                checkResult.onSuccess { status ->
                    result.success(Converters.permissionStatusToString(status))
                }
                checkResult.onFailure { error, _ ->
                    result.error("PERMISSION_ERROR", "Permission check failed: ${error.description}", null)
                }
            } catch (e: Exception) {
                result.error("PERMISSION_ERROR", "Failed to check permission: ${e.message}", null)
            }
        }
    }

    private fun handleRequestPermission(call: MethodCall, result: MethodChannel.Result) {
        val act = activity
        val contract = permissionContract
        if (act == null || contract == null) {
            result.error("NO_ACTIVITY", "Activity not available for permission request", null)
            return
        }
        if (pendingPermissionResult != null) {
            result.error("IN_PROGRESS", "A permission request is already in progress", null)
            return
        }
        val permissionName = call.argument<String>("permission") ?: "camera"
        val permission = when (permissionName) {
            "camera" -> Permission.CAMERA
            else -> Permission.CAMERA
        }

        // Check if the contract can resolve synchronously (permission already granted/denied)
        val syncResult = contract.getSynchronousResult(act, permission)
        if (syncResult != null) {
            val datResult = syncResult.value
            datResult.onSuccess { status ->
                result.success(Converters.permissionStatusToString(status))
            }
            datResult.onFailure { error, _ ->
                result.error("PERMISSION_ERROR", "Permission request failed: ${error.description}", null)
            }
            return
        }

        // Need to launch the Meta AI app for permission - use startActivityForResult
        pendingPermissionResult = result
        val intent = contract.createIntent(act, permission)
        act.startActivityForResult(intent, PERMISSION_REQUEST_CODE)
    }

    fun handlePermissionResult(datResult: DatResult<PermissionStatus, PermissionError>) {
        val result = pendingPermissionResult ?: return
        pendingPermissionResult = null
        datResult.onSuccess { status ->
            result.success(Converters.permissionStatusToString(status))
        }
        datResult.onFailure { error, _ ->
            result.error("PERMISSION_ERROR", "Permission request failed: ${error.description}", null)
        }
    }

    private fun handleUrl(call: MethodCall, result: MethodChannel.Result) {
        // handleUrl is primarily for iOS deep link callbacks
        result.success(false)
    }

    private fun handleStartStreamSession(call: MethodCall, result: MethodChannel.Result) {
        val app = application
        val registry = textureRegistry
        if (app == null || registry == null) {
            result.error("NO_CONTEXT", "Application or texture registry not available", null)
            return
        }
        try {
            val qualityStr = call.argument<String>("videoQuality") ?: "medium"
            val frameRate = call.argument<Int>("frameRate") ?: 24
            val videoQuality = Converters.videoQualityFromString(qualityStr)

            val manager = StreamSessionManager(
                application = app,
                textureRegistry = registry,
                scope = scope,
                streamStateHandler = streamStateHandler,
                videoFrameInfoHandler = videoFrameInfoHandler,
                streamErrorsHandler = streamErrorsHandler
            )
            streamSessionManager = manager

            val sessionInfo = manager.startSession(videoQuality, frameRate)
            result.success(sessionInfo)
        } catch (e: Exception) {
            result.error("STREAM_ERROR", "Failed to start stream session: ${e.message}", null)
        }
    }

    private fun handleStopStreamSession(result: MethodChannel.Result) {
        try {
            streamSessionManager?.stopSession()
            streamSessionManager = null
            result.success(null)
        } catch (e: Exception) {
            result.error("STREAM_ERROR", "Failed to stop stream session: ${e.message}", null)
        }
    }

    private fun handleCapturePhoto(result: MethodChannel.Result) {
        val manager = streamSessionManager
        if (manager == null) {
            result.error("NO_SESSION", "No active stream session", null)
            return
        }
        scope.launch(Dispatchers.Main) {
            try {
                val photoResult = manager.capturePhoto()
                if (photoResult != null) {
                    result.success(photoResult)
                } else {
                    result.error("CAPTURE_ERROR", "Failed to capture photo", null)
                }
            } catch (e: Exception) {
                result.error("CAPTURE_ERROR", "Photo capture failed: ${e.message}", null)
            }
        }
    }

    private fun handleDispose(result: MethodChannel.Result) {
        streamSessionManager?.stopSession()
        streamSessionManager = null
        mockDeviceKitHandler?.dispose()
        mockDeviceKitHandler = null
        result.success(null)
    }

    private fun getMockHandler(): MockDeviceKitHandler? {
        val app = application ?: return null
        if (mockDeviceKitHandler == null) {
            mockDeviceKitHandler = MockDeviceKitHandler(app)
        }
        return mockDeviceKitHandler
    }

    fun dispose() {
        streamSessionManager?.stopSession()
        streamSessionManager = null
        mockDeviceKitHandler?.dispose()
        mockDeviceKitHandler = null
    }
}
